<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Cleaning and exploring NCBI data with the `bioseq` package â€¢ bioseq</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Cleaning and exploring NCBI data with the `bioseq` package">
<meta property="og:description" content="bioseq">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">bioseq</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/intro-bioseq.html">Introduction to the `bioseq` package</a>
    </li>
    <li>
      <a href="../articles/ref_database.html">Cleaning and exploring NCBI data with the `bioseq` package</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/fkeck/bioseq/">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Cleaning and exploring NCBI data with the <code>bioseq</code> package</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/fkeck/bioseq/blob/master/vignettes/ref_database.Rmd"><code>vignettes/ref_database.Rmd</code></a></small>
      <div class="hidden name"><code>ref_database.Rmd</code></div>

    </div>

    
    
<p>The purpose of this vignette is to illustrate the use of the <code>bioseq</code> package with a practical case. For a general introduction to the package please see the vignette <em>Introduction to the bioseq package</em>. Here we show how the functions of <code>bioseq</code> can be used to clean and explore real biological data. We take as example a set of rbcL sequences of Fragilaria species retrieved from NCBI (these data ared distributed with the package, see <code><a href="../reference/fragilaria.html">?fragilaria</a></code>). Fragilaria is a genus of freshwater and saltwater diatoms. Diatoms are a major group of microscopic algae.</p>
<p>Data obtained from generic repository like NCBI or BOLD often need to be filtered and cleaned before they can be used in a project. Here we show how this can be accomplished using <code>bioseq</code> and more particularly how to perform some basic cleaning, extract a specific barcode region from raw sequences, and use functions from other packages. Some of the code chunks presented here can be adapted to reconstruct a phylogeny for a particular taxonomic group or to prepare a reference database against which metabarcoding sequencing reads will be matched for taxonomic identification.</p>
<div class="figure">
<img src="../../../../../Google%20Drive/Sync%20work/packages/bioseq/vignettes/f_pararumpens.png" alt=""><p class="caption">Fragilaria pararumpens. Photo by Dr Maria Kahlert</p>
</div>
<div id="loading-packages" class="section level2">
<h2 class="hasAnchor">
<a href="#loading-packages" class="anchor"></a>Loading packages</h2>
<p>We start by loading the package <code>bioseq</code>. It is assumed that you already intalled the package either from CRAN using the function <code>install.packages("bioseq)</code> or from GitHub using <code><a href="https://rdrr.io/pkg/remotes/man/install_github.html">remotes::install_github("fkeck/bioseq")</a></code>.</p>
<div class="sourceCode" id="cb1"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">bioseq</span>)</pre></body></html></div>
<p>We additionaly load the <code>tidyverse</code> package. This is a meta-package which loads several useful packages that we will need for later. In particular we will use <code>tibble</code> and <code>dplyr</code> to work with tibbles. This allows to work with other types of data and to centralize everything in one place. To learn more about the tidyverse and these packages you can visit.</p>
<div class="sourceCode" id="cb2"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">tidyverse</span>)</pre></body></html></div>
</div>
<div id="loading-data" class="section level2">
<h2 class="hasAnchor">
<a href="#loading-data" class="anchor"></a>Loading data</h2>
<p>Genetic data are often delivered in FASTA format. In this example we will use rbcL sequences of Fragilaria species retrieved from NCBI and directly available as unparsed FASTA in the package. We can use the function <code>read_fasta</code> to directly read a FASTA file into R. Here we read a vector of character already available in R, but most of the time the argument <code>file</code> will be used to pass a path to a file available locally or via the network.</p>
<div class="sourceCode" id="cb3"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span>(<span class="no">fragilaria</span>, <span class="kw">package</span> <span class="kw">=</span> <span class="st">"bioseq"</span>)
<span class="no">fra_data</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/read_fasta.html">read_fasta</a></span>(<span class="no">fragilaria</span>)

<span class="no">fra_data</span>
<span class="co">#&gt; DNA vector of 41 sequences</span>
<span class="co">#&gt; MK576039.1 Fragil...  TTCATCGCTTATGAATGTGATTTATTTGAAGAAGGTTCTTTAGCTAACTTAACAGCTTCT... + 575 bases</span>
<span class="co">#&gt; MK672908.1 Fragil...  GCTTTCATCGCTTACGAATGTGATTTATTTGAAGAAGGTTCTTTAGCTAACTTAACAGCT... + 1107 bases</span>
<span class="co">#&gt; MK672907.1 Fragil...  ATGTCTCAATCTGTATCAGAACGGACTCGAATCAAAAGTGACCGTTACGAATCTGGTGTA... + 1386 bases</span>
<span class="co">#&gt; MK672906.1 Fragil...  ATGTCTCAATCTGTATCAGAACGGACTCGAATCAAAAGTGACCGTTACGAATCTGGTGTA... + 1386 bases</span>
<span class="co">#&gt; KT072939.1 Fragil...  TTAGCATTATTCCGTATTACTCCACAACCAGGTGTAGATCCAGTAGAAGCAGCAGCAGCA... + 744 bases</span>
<span class="co">#&gt; KT072928.1 Fragil...  TTAGCATTATTCCGTATTACTCCACAACCAGGTGTAGATCCAGTAGAAGCAGCAGCAGCA... + 744 bases</span>
<span class="co">#&gt; KT072903.1 Fragil...  TTAGCATTATTCCGTATTACTCCACAACCAGGTGTAGATCCAGTAGAAGCAGCAGCAGCA... + 744 bases</span>
<span class="co">#&gt; MF092945.1 Fragil...  AGTGACCGTTACGAATCTGGTGTAATCCCTTACGCTAAAATGGGTTACTGGGATGCTTCA... + 1263 bases</span>
<span class="co">#&gt; MF092943.1 Fragil...  AGTGACCGTTACGAATCTGGTGTAATCCCTTACGCTAAAATGGGTTACTGGGATGCTTCA... + 1263 bases</span>
<span class="co">#&gt; JQ003574.1 Synedr...  AAAAANTATGGGCGTGTAGTATACGAAGGTTTAAAAGGTGGATTAGACTTCTTAAAAGAT... + 390 bases</span>
<span class="co">#&gt; JQ003570.1 Fragil...  ATTCCACACTCATACTTAAAAACATTCCAAGGTCCAGCAACTGGTATCATCGTAGAACGT... + 516 bases</span>
<span class="co">#&gt; JQ003569.1 Synedr...  AAAAACTACGGTCGTGTAGTATATGAAGGTTTAAAAGGTGGTTTAGACTTCTTAAAAGAT... + 390 bases</span>
<span class="co">#&gt; ... with 29 more sequences.</span></pre></body></html></div>
<p>According to the output, we have a DNA vector with 41 sequences. Those sequences were retrieved directly from NCBI. Therefore, they are not aligned and they have different lenght. It is easy to get the range of lengths with the following command:</p>
<div class="sourceCode" id="cb4"><html><body><pre class="r"><span class="fu"><a href="../reference/seq_nchar.html">seq_nchar</a></span>(<span class="no">fra_data</span>) <span class="kw">%&gt;%</span> <span class="fu"><a href="https://rdrr.io/r/base/range.html">range</a></span>()
<span class="co">#&gt; [1]  450 1510</span></pre></body></html></div>
<p>As explained in the introduction, we want to analyze the data in a dataframe as it present several advantages. The <code>bioseq</code> package has been designed to work smoothly with <code>tibbles</code>, an extension of the standard dataframe. Below, we use the <code>as_tibble</code> function to turn the DNA vector into a <code>tibble</code> with two columns (one for the names and one for the sequences). We will also clean the names and split them into two columns. Finally we create an extra column with the length (number of characters) of each sequence.</p>
<div class="sourceCode" id="cb5"><html><body><pre class="r"><span class="no">fra_data</span> <span class="kw">&lt;-</span> <span class="fu">tibble</span>(<span class="kw">label</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span>(<span class="no">fra_data</span>), <span class="kw">sequence</span> <span class="kw">=</span> <span class="no">fra_data</span>)</pre></body></html></div>
<div class="sourceCode" id="cb6"><html><body><pre class="r"><span class="no">fra_data</span> <span class="kw">&lt;-</span> <span class="no">fra_data</span> <span class="kw">%&gt;%</span>
  <span class="fu">mutate</span>(<span class="kw">genbank_id</span> <span class="kw">=</span> <span class="fu">str_extract</span>(<span class="no">label</span>, <span class="st">"([^\\s]+)"</span>),
         <span class="kw">taxa</span> <span class="kw">=</span> <span class="fu">str_extract</span>(<span class="no">label</span>, <span class="st">"(?&lt;= ).*"</span>)) <span class="kw">%&gt;%</span>
  <span class="fu">select</span>(<span class="no">genbank_id</span>, <span class="no">taxa</span>, <span class="no">sequence</span>)

<span class="no">fra_data</span> <span class="kw">&lt;-</span> <span class="no">fra_data</span> <span class="kw">%&gt;%</span>
  <span class="fu">mutate</span>(<span class="kw">n_base</span> <span class="kw">=</span> <span class="fu"><a href="../reference/seq_nchar.html">seq_nchar</a></span>(<span class="no">sequence</span>))

<span class="no">fra_data</span>
<span class="co">#&gt; # A tibble: 41 x 4</span>
<span class="co">#&gt;    genbank_id taxa           sequence                                     n_base</span>
<span class="co">#&gt;    &lt;chr&gt;      &lt;chr&gt;          &lt;DNA&gt;                                         &lt;int&gt;</span>
<span class="co">#&gt;  1 MK576039.1 Fragilaria MKâ€¦ TTCATCGCTTATGAATGTGATTTATTTGAAGAAGGTTCTTTAGâ€¦    635</span>
<span class="co">#&gt;  2 MK672908.1 Fragilaria joâ€¦ GCTTTCATCGCTTACGAATGTGATTTATTTGAAGAAGGTTCTTâ€¦   1167</span>
<span class="co">#&gt;  3 MK672907.1 Fragilaria heâ€¦ ATGTCTCAATCTGTATCAGAACGGACTCGAATCAAAAGTGACCâ€¦   1446</span>
<span class="co">#&gt;  4 MK672906.1 Fragilaria agâ€¦ ATGTCTCAATCTGTATCAGAACGGACTCGAATCAAAAGTGACCâ€¦   1446</span>
<span class="co">#&gt;  5 KT072939.1 Fragilaria ruâ€¦ TTAGCATTATTCCGTATTACTCCACAACCAGGTGTAGATCCAGâ€¦    804</span>
<span class="co">#&gt;  6 KT072928.1 Fragilaria caâ€¦ TTAGCATTATTCCGTATTACTCCACAACCAGGTGTAGATCCAGâ€¦    804</span>
<span class="co">#&gt;  7 KT072903.1 Fragilaria crâ€¦ TTAGCATTATTCCGTATTACTCCACAACCAGGTGTAGATCCAGâ€¦    804</span>
<span class="co">#&gt;  8 MF092945.1 Fragilaria MFâ€¦ AGTGACCGTTACGAATCTGGTGTAATCCCTTACGCTAAAATGGâ€¦   1323</span>
<span class="co">#&gt;  9 MF092943.1 Fragilaria MFâ€¦ AGTGACCGTTACGAATCTGGTGTAATCCCTTACGCTAAAATGGâ€¦   1323</span>
<span class="co">#&gt; 10 JQ003574.1 Synedra radiaâ€¦ AAAAANTATGGGCGTGTAGTATACGAAGGTTTAAAAGGTGGATâ€¦    450</span>
<span class="co">#&gt; # â€¦ with 31 more rows</span></pre></body></html></div>
<div id="cropping-sequences" class="section level3">
<h3 class="hasAnchor">
<a href="#cropping-sequences" class="anchor"></a>Cropping sequences</h3>
<p>In this example we are only interested in a particular region of the sequences, a short fragment used in many diatom metabarcoding studies. To extract the region of interest we can provide two nucleotide position but this solution require the sequences to be aligned. The alternative solution is to crop using nucleotide patterns. Here, we will use the forward and reverse primers delimiting the barcode region to crop the sequences. First, we create two DNA vectors one for the forward and one for the reverse primers:</p>
<div class="sourceCode" id="cb7"><html><body><pre class="r"><span class="no">FWD</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/dna.html">dna</a></span>(<span class="st">"AGGTGAAGTAAAAGGTTCWTACTTAAA"</span>,
           <span class="st">"AGGTGAAGTTAAAGGTTCWTAYTTAAA"</span>,
           <span class="st">"AGGTGAAACTAAAGGTTCWTACTTAAA"</span>)

<span class="no">REV</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/dna.html">dna</a></span>(<span class="st">"CAGTWGTWGGTAAATTAGAAGG"</span>,
           <span class="st">"CTGTTGTWGGTAAATTAGAAGG"</span>)</pre></body></html></div>
<p>You may have noticed that some characters of the primers are ambiguous which means they can represent several nucleotides. To get all the combinations represented by those ambiguous primer sequences we can use the <code><a href="../reference/seq_disambiguate_IUPAC.html">seq_disambiguate_IUPAC()</a></code> function. For example for the forward primers:</p>
<div class="sourceCode" id="cb8"><html><body><pre class="r"><span class="fu"><a href="../reference/seq_disambiguate_IUPAC.html">seq_disambiguate_IUPAC</a></span>(<span class="no">FWD</span>)
<span class="co">#&gt; [[1]]</span>
<span class="co">#&gt; DNA vector of 2 sequences</span>
<span class="co">#&gt; &gt;   AGGTGAAGTAAAAGGTTCATACTTAAA</span>
<span class="co">#&gt; &gt;   AGGTGAAGTAAAAGGTTCTTACTTAAA</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; [[2]]</span>
<span class="co">#&gt; DNA vector of 4 sequences</span>
<span class="co">#&gt; &gt;   AGGTGAAGTTAAAGGTTCATACTTAAA</span>
<span class="co">#&gt; &gt;   AGGTGAAGTTAAAGGTTCTTACTTAAA</span>
<span class="co">#&gt; &gt;   AGGTGAAGTTAAAGGTTCATATTTAAA</span>
<span class="co">#&gt; &gt;   AGGTGAAGTTAAAGGTTCTTATTTAAA</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; [[3]]</span>
<span class="co">#&gt; DNA vector of 2 sequences</span>
<span class="co">#&gt; &gt;   AGGTGAAACTAAAGGTTCATACTTAAA</span>
<span class="co">#&gt; &gt;   AGGTGAAACTAAAGGTTCTTACTTAAA</span></pre></body></html></div>
<p>We can see that the three ambiguous sequences used as forward primers actually represent eight different sequences. Fortunately, the functions of <code>bioseq</code> can automatically recognize ambiguous nucleotides and interpret them as such. This is why it is always recommended to construct DNA/RNA/AA vectors instead of simple character vector when it is possible.</p>
<p>Now we crop the sequences using the forward and reverse primers as delimiters. We can expect that the returned sequences will be aligned because they correspond to a very specific region with a fixed length.</p>
<div class="sourceCode" id="cb9"><html><body><pre class="r"><span class="no">fra_data</span> <span class="kw">&lt;-</span> <span class="no">fra_data</span> <span class="kw">%&gt;%</span>
  <span class="fu">mutate</span>(<span class="kw">barcode</span> <span class="kw">=</span> <span class="fu"><a href="../reference/seq_crop_pattern.html">seq_crop_pattern</a></span>(<span class="no">sequence</span>,
                                    <span class="kw">pattern_in</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span>(<span class="no">FWD</span>),
                                    <span class="kw">pattern_out</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span>(<span class="no">REV</span>)))

<span class="no">fra_data</span>
<span class="co">#&gt; # A tibble: 41 x 5</span>
<span class="co">#&gt;    genbank_id taxa       sequence                 n_base barcode                </span>
<span class="co">#&gt;    &lt;chr&gt;      &lt;chr&gt;      &lt;DNA&gt;                     &lt;int&gt; &lt;DNA&gt;                  </span>
<span class="co">#&gt;  1 MK576039.1 Fragilariâ€¦ TTCATCGCTTATGAATGTGATTTâ€¦    635 NA                     </span>
<span class="co">#&gt;  2 MK672908.1 Fragilariâ€¦ GCTTTCATCGCTTACGAATGTGAâ€¦   1167 CATCACTGCTGCTACTATGGAAâ€¦</span>
<span class="co">#&gt;  3 MK672907.1 Fragilariâ€¦ ATGTCTCAATCTGTATCAGAACGâ€¦   1446 CATCACTGCTGCTACTATGGAAâ€¦</span>
<span class="co">#&gt;  4 MK672906.1 Fragilariâ€¦ ATGTCTCAATCTGTATCAGAACGâ€¦   1446 CATCACTGCTGCTACTATGGAAâ€¦</span>
<span class="co">#&gt;  5 KT072939.1 Fragilariâ€¦ TTAGCATTATTCCGTATTACTCCâ€¦    804 NA                     </span>
<span class="co">#&gt;  6 KT072928.1 Fragilariâ€¦ TTAGCATTATTCCGTATTACTCCâ€¦    804 NA                     </span>
<span class="co">#&gt;  7 KT072903.1 Fragilariâ€¦ TTAGCATTATTCCGTATTACTCCâ€¦    804 NA                     </span>
<span class="co">#&gt;  8 MF092945.1 Fragilariâ€¦ AGTGACCGTTACGAATCTGGTGTâ€¦   1323 CATCACAGCTGCTACTATGGAAâ€¦</span>
<span class="co">#&gt;  9 MF092943.1 Fragilariâ€¦ AGTGACCGTTACGAATCTGGTGTâ€¦   1323 CATTACTGCTGCTACTATGGATâ€¦</span>
<span class="co">#&gt; 10 JQ003574.1 Synedra râ€¦ AAAAANTATGGGCGTGTAGTATAâ€¦    450 NA                     </span>
<span class="co">#&gt; # â€¦ with 31 more rows</span></pre></body></html></div>
<p>As expected, the barcode sequences seem to be aligned. We can also see that this operation has produced several NAs which correspond to sequences where the primers were not detected, probably because the barcode region was not included in the sequence. We could simply exclude the NA values, but it is safer to filter only the sequences whose length match exactly the expected length of the barcode (i.e.Â 263 bp).</p>
<div class="sourceCode" id="cb10"><html><body><pre class="r"><span class="no">fra_data</span> <span class="kw">&lt;-</span> <span class="no">fra_data</span> <span class="kw">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/r/stats/filter.html">filter</a></span>(<span class="fu"><a href="../reference/seq_nchar.html">seq_nchar</a></span>(<span class="no">barcode</span>) <span class="kw">==</span> <span class="fl">263</span>)

<span class="no">fra_data</span>
<span class="co">#&gt; # A tibble: 23 x 5</span>
<span class="co">#&gt;    genbank_id taxa      sequence                  n_base barcode                </span>
<span class="co">#&gt;    &lt;chr&gt;      &lt;chr&gt;     &lt;DNA&gt;                      &lt;int&gt; &lt;DNA&gt;                  </span>
<span class="co">#&gt;  1 MK672908.1 Fragilarâ€¦ GCTTTCATCGCTTACGAATGTGATâ€¦   1167 CATCACTGCTGCTACTATGGAAâ€¦</span>
<span class="co">#&gt;  2 MK672907.1 Fragilarâ€¦ ATGTCTCAATCTGTATCAGAACGGâ€¦   1446 CATCACTGCTGCTACTATGGAAâ€¦</span>
<span class="co">#&gt;  3 MK672906.1 Fragilarâ€¦ ATGTCTCAATCTGTATCAGAACGGâ€¦   1446 CATCACTGCTGCTACTATGGAAâ€¦</span>
<span class="co">#&gt;  4 MF092945.1 Fragilarâ€¦ AGTGACCGTTACGAATCTGGTGTAâ€¦   1323 CATCACAGCTGCTACTATGGAAâ€¦</span>
<span class="co">#&gt;  5 MF092943.1 Fragilarâ€¦ AGTGACCGTTACGAATCTGGTGTAâ€¦   1323 CATTACTGCTGCTACTATGGATâ€¦</span>
<span class="co">#&gt;  6 EU090046.1 Fragilarâ€¦ TGACCGTTACGAATCTGGTGTAATâ€¦   1330 CATCACAGCTGCTACTATGGAAâ€¦</span>
<span class="co">#&gt;  7 EU090045.1 Fragilarâ€¦ TGACCGTTACGAATCTGGTGTAATâ€¦   1330 CATCACAGCTGCTACTATGGAAâ€¦</span>
<span class="co">#&gt;  8 EU090044.1 Fragilarâ€¦ TGACCGTTACGAATCTGGTGTAATâ€¦   1330 CATCACAGCTGCTACTATGGAAâ€¦</span>
<span class="co">#&gt;  9 EU090043.1 Fragilarâ€¦ TGACCGTTACGAATCTGGTGTAATâ€¦   1328 CATCACAGCTGCTACTATGGAAâ€¦</span>
<span class="co">#&gt; 10 EU090042.1 Fragilarâ€¦ ACCGTTACGAATCTGGTGTAATCCâ€¦   1329 CATCACAGCTGCTACTATGGAAâ€¦</span>
<span class="co">#&gt; # â€¦ with 13 more rows</span></pre></body></html></div>
</div>
<div id="consensus-sequences-and-phylogeny" class="section level3">
<h3 class="hasAnchor">
<a href="#consensus-sequences-and-phylogeny" class="anchor"></a>Consensus sequences and phylogeny</h3>
<p>We now want to reconstruct a phylogenetic tree for the sequences. We can include all the sequences in the tree but for clarity we want keep only one sequence per taxa. Instead of selecting one sequence randomly we compute a consensus sequence for each taxa using the function <code><a href="../reference/seq_consensus.html">seq_consensus()</a></code>. There are different methods to compute a consensus sequence implemented in the function, you can check the manual for details. The default method used below uses the majority rule.</p>
<div class="sourceCode" id="cb11"><html><body><pre class="r"><span class="no">fra_consensus</span> <span class="kw">&lt;-</span> <span class="no">fra_data</span> <span class="kw">%&gt;%</span>
  <span class="fu">group_by</span>(<span class="no">taxa</span>) <span class="kw">%&gt;%</span>
  <span class="fu">summarise</span>(<span class="kw">consensus_barcode</span> <span class="kw">=</span> <span class="fu"><a href="../reference/seq_consensus.html">seq_consensus</a></span>(<span class="no">barcode</span>))

<span class="no">fra_consensus</span>
<span class="co">#&gt; # A tibble: 20 x 2</span>
<span class="co">#&gt;    taxa              consensus_barcode                                          </span>
<span class="co">#&gt;    &lt;chr&gt;             &lt;DNA&gt;                                                      </span>
<span class="co">#&gt;  1 Fragilaria agnesâ€¦ CATCACTGCTGCTACTATGGAAGAACTTTACAAACGTGGTGAGTACGCTAAAGCTGTAâ€¦</span>
<span class="co">#&gt;  2 Fragilaria bidens CATCACTGCTGCTACTATGGAAGAACTTTACAAACGTGGTGATTACGCTAAAGCTGTAâ€¦</span>
<span class="co">#&gt;  3 Fragilaria crotoâ€¦ CATCACTGCTGCTACTATGGAAGAACTTTACAAACGTGGTGCTTACGCTAAAGCTGTAâ€¦</span>
<span class="co">#&gt;  4 Fragilaria EU090â€¦ CATCACAGCTGCTACTATGGAAGAAGTTTACAAGCGTGGTGAATATGCCAAAGCTGTAâ€¦</span>
<span class="co">#&gt;  5 Fragilaria EU090â€¦ CATCACAGCTGCTACTATGGAAGAAGTTTACAAGCGTGGTGAATATGCTAAAGCTGTAâ€¦</span>
<span class="co">#&gt;  6 Fragilaria EU090â€¦ CATCACAGCTGCTACTATGGAAGAAGTTTACAAGCGTGGTGAATATGCTAAAGCTGTAâ€¦</span>
<span class="co">#&gt;  7 Fragilaria EU090â€¦ CATCACAGCTGCTACTATGGAAGAAGTTTACAAGCGTGGTGAATATGCTAAAGCTGTAâ€¦</span>
<span class="co">#&gt;  8 Fragilaria EU090â€¦ CATCACAGCTGCTACTATGGAAGAAGTTTACAAGCGTGGTGAATATGCTAAAGCTGTAâ€¦</span>
<span class="co">#&gt;  9 Fragilaria EU090â€¦ CATCACAGCTGCTACTATGGAAGAAGTTTACAAGCGTGGTGAATATGCTAAAGCTGTAâ€¦</span>
<span class="co">#&gt; 10 Fragilaria EU090â€¦ CATCACAGCTGCTACTATGGAAGAAGTTTACAAGCGTGGTGAATATGCTAAAGCTGTAâ€¦</span>
<span class="co">#&gt; 11 Fragilaria famelâ€¦ CATCACTGCTGCTACTATGGAAGAAGTTTACAAACGTGGTGAGTATGCTAAAGCTGTAâ€¦</span>
<span class="co">#&gt; 12 Fragilaria heathâ€¦ CATCACTGCTGCTACTATGGAAGAGCTTTACAAACGTGGTGAGTACGCTAAAGCTGTAâ€¦</span>
<span class="co">#&gt; 13 Fragilaria joachâ€¦ CATCACTGCTGCTACTATGGAAGAACTTTACAAACGTGGTGAGTACGCTAAAGCTGTAâ€¦</span>
<span class="co">#&gt; 14 Fragilaria KC969â€¦ CATCACAGCTGCTACTATGGAAGAAGTTTACAAGCGTGGTGAATATGCCAAAGCTGTAâ€¦</span>
<span class="co">#&gt; 15 Fragilaria KC969â€¦ CGTTACAGCTGCTACTATGGAAGATGTATACAAACGTAGTGAGTATGCTAAAGCTGTAâ€¦</span>
<span class="co">#&gt; 16 Fragilaria MF092â€¦ CATTACTGCTGCTACTATGGATGAACTTTACAAACGTGGTGAGTACGCTAAAGCTGTAâ€¦</span>
<span class="co">#&gt; 17 Fragilaria MF092â€¦ CATCACAGCTGCTACTATGGAAGAACTTTACAAACGTGGTGAATACGCTAAAGCTGTAâ€¦</span>
<span class="co">#&gt; 18 Fragilaria permiâ€¦ CATTACTGCTGCTACTATGGAAGAACTTTACAAACGTGGTGAGTACGCTAAAGCTGTAâ€¦</span>
<span class="co">#&gt; 19 Fragilaria rumpeâ€¦ CATCACGGCTGCTACTATGGAAGAACTTTACAAACGTGGTGCTTACGCTAAAGCTGTAâ€¦</span>
<span class="co">#&gt; 20 Fragilaria striaâ€¦ CATCACAGCTGCTACTATGGAAGAAGTTTACAAGCGTGGTGAATATGCCAAAGCTGTAâ€¦</span></pre></body></html></div>
<p>Next, we will use <code>ape</code> to do a quick phylogenetic reconstruction of these sequences. We can use the function <code>as_DNAbin</code> to convert a tibble into a <code>DNAbin</code> object and then use the functions of <code>ape</code> to reconstruct and plot a phylogenetic tree with the distance-based BIONJ algorithm.</p>
<div class="sourceCode" id="cb12"><html><body><pre class="r"><span class="no">fra_consensus</span> <span class="kw">%&gt;%</span>
  <span class="fu"><a href="../reference/as_DNAbin.html">as_DNAbin</a></span>(<span class="no">consensus_barcode</span>, <span class="no">taxa</span>) <span class="kw">%&gt;%</span>
  <span class="kw pkg">ape</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/ape/man/dist.dna.html">dist.dna</a></span>() <span class="kw">%&gt;%</span>
  <span class="kw pkg">ape</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/ape/man/bionj.html">bionj</a></span>() <span class="kw">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/r/graphics/plot.html">plot</a></span>()</pre></body></html></div>
<p><img src="ref_database_files/figure-html/unnamed-chunk-14-1.png" width="576"></p>
</div>
<div id="clustering-sequences" class="section level3">
<h3 class="hasAnchor">
<a href="#clustering-sequences" class="anchor"></a>Clustering sequences</h3>
<p>The tree above suggests that several sequences are exactly the same. This can be quickly assessed:</p>
<div class="sourceCode" id="cb13"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/duplicated.html">duplicated</a></span>(<span class="no">fra_consensus</span>$<span class="no">consensus_barcode</span>)
<span class="co">#&gt;  [1] FALSE FALSE FALSE FALSE FALSE  TRUE  TRUE  TRUE  TRUE  TRUE FALSE FALSE</span>
<span class="co">#&gt; [13] FALSE  TRUE FALSE FALSE FALSE FALSE FALSE  TRUE</span></pre></body></html></div>
<p>Indeed, several sequences are duplicates and we may want to keep only unique sequences. We could select only unique sequence but here we rather use the <code>seq_cluster</code> function to cluster sequences based on their similarity.</p>
<div class="sourceCode" id="cb14"><html><body><pre class="r"><span class="no">fra_consensus</span> <span class="kw">&lt;-</span>
  <span class="no">fra_consensus</span> <span class="kw">%&gt;%</span>
  <span class="fu">mutate</span>(<span class="kw">cluster</span> <span class="kw">=</span> <span class="fu"><a href="../reference/seq_cluster.html">seq_cluster</a></span>(<span class="no">consensus_barcode</span>,
                               <span class="kw">threshold</span> <span class="kw">=</span> <span class="fl">0.001</span>))
<span class="no">fra_consensus</span>
<span class="co">#&gt; # A tibble: 20 x 3</span>
<span class="co">#&gt;    taxa             consensus_barcode                                    cluster</span>
<span class="co">#&gt;    &lt;chr&gt;            &lt;DNA&gt;                                                  &lt;int&gt;</span>
<span class="co">#&gt;  1 Fragilaria agneâ€¦ CATCACTGCTGCTACTATGGAAGAACTTTACAAACGTGGTGAGTACGCTAAâ€¦       1</span>
<span class="co">#&gt;  2 Fragilaria bideâ€¦ CATCACTGCTGCTACTATGGAAGAACTTTACAAACGTGGTGATTACGCTAAâ€¦       2</span>
<span class="co">#&gt;  3 Fragilaria crotâ€¦ CATCACTGCTGCTACTATGGAAGAACTTTACAAACGTGGTGCTTACGCTAAâ€¦       3</span>
<span class="co">#&gt;  4 Fragilaria EU09â€¦ CATCACAGCTGCTACTATGGAAGAAGTTTACAAGCGTGGTGAATATGCCAAâ€¦       4</span>
<span class="co">#&gt;  5 Fragilaria EU09â€¦ CATCACAGCTGCTACTATGGAAGAAGTTTACAAGCGTGGTGAATATGCTAAâ€¦       5</span>
<span class="co">#&gt;  6 Fragilaria EU09â€¦ CATCACAGCTGCTACTATGGAAGAAGTTTACAAGCGTGGTGAATATGCTAAâ€¦       5</span>
<span class="co">#&gt;  7 Fragilaria EU09â€¦ CATCACAGCTGCTACTATGGAAGAAGTTTACAAGCGTGGTGAATATGCTAAâ€¦       5</span>
<span class="co">#&gt;  8 Fragilaria EU09â€¦ CATCACAGCTGCTACTATGGAAGAAGTTTACAAGCGTGGTGAATATGCTAAâ€¦       5</span>
<span class="co">#&gt;  9 Fragilaria EU09â€¦ CATCACAGCTGCTACTATGGAAGAAGTTTACAAGCGTGGTGAATATGCTAAâ€¦       5</span>
<span class="co">#&gt; 10 Fragilaria EU09â€¦ CATCACAGCTGCTACTATGGAAGAAGTTTACAAGCGTGGTGAATATGCTAAâ€¦       5</span>
<span class="co">#&gt; 11 Fragilaria fameâ€¦ CATCACTGCTGCTACTATGGAAGAAGTTTACAAACGTGGTGAGTATGCTAAâ€¦       6</span>
<span class="co">#&gt; 12 Fragilaria heatâ€¦ CATCACTGCTGCTACTATGGAAGAGCTTTACAAACGTGGTGAGTACGCTAAâ€¦       7</span>
<span class="co">#&gt; 13 Fragilaria joacâ€¦ CATCACTGCTGCTACTATGGAAGAACTTTACAAACGTGGTGAGTACGCTAAâ€¦       8</span>
<span class="co">#&gt; 14 Fragilaria KC96â€¦ CATCACAGCTGCTACTATGGAAGAAGTTTACAAGCGTGGTGAATATGCCAAâ€¦       4</span>
<span class="co">#&gt; 15 Fragilaria KC96â€¦ CGTTACAGCTGCTACTATGGAAGATGTATACAAACGTAGTGAGTATGCTAAâ€¦       9</span>
<span class="co">#&gt; 16 Fragilaria MF09â€¦ CATTACTGCTGCTACTATGGATGAACTTTACAAACGTGGTGAGTACGCTAAâ€¦      10</span>
<span class="co">#&gt; 17 Fragilaria MF09â€¦ CATCACAGCTGCTACTATGGAAGAACTTTACAAACGTGGTGAATACGCTAAâ€¦      11</span>
<span class="co">#&gt; 18 Fragilaria permâ€¦ CATTACTGCTGCTACTATGGAAGAACTTTACAAACGTGGTGAGTACGCTAAâ€¦      12</span>
<span class="co">#&gt; 19 Fragilaria rumpâ€¦ CATCACGGCTGCTACTATGGAAGAACTTTACAAACGTGGTGCTTACGCTAAâ€¦      13</span>
<span class="co">#&gt; 20 Fragilaria striâ€¦ CATCACAGCTGCTACTATGGAAGAAGTTTACAAGCGTGGTGAATATGCCAAâ€¦       4</span></pre></body></html></div>
<p>Finally we recompute a consensus sequence for each cluster and we reconstruct the phylogenetic tree.</p>
<div class="sourceCode" id="cb15"><html><body><pre class="r"><span class="no">fra_consensus</span> <span class="kw">&lt;-</span>
  <span class="no">fra_consensus</span> <span class="kw">%&gt;%</span>
  <span class="fu">group_by</span>(<span class="no">cluster</span>) <span class="kw">%&gt;%</span>
  <span class="fu">summarise</span>(<span class="kw">taxa_group</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste</a></span>(<span class="no">taxa</span>, <span class="kw">collapse</span> <span class="kw">=</span> <span class="st">"/"</span>),
            <span class="kw">consensus_barcode</span> <span class="kw">=</span> <span class="fu"><a href="../reference/seq_consensus.html">seq_consensus</a></span>(<span class="no">consensus_barcode</span>))

<span class="no">fra_consensus</span> <span class="kw">%&gt;%</span>
  <span class="fu"><a href="../reference/as_DNAbin.html">as_DNAbin</a></span>(<span class="no">consensus_barcode</span>, <span class="no">taxa_group</span>) <span class="kw">%&gt;%</span>
  <span class="kw pkg">ape</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/ape/man/dist.dna.html">dist.dna</a></span>() <span class="kw">%&gt;%</span>
  <span class="kw pkg">ape</span><span class="kw ns">::</span><span class="fu"><a href="https://rdrr.io/pkg/ape/man/bionj.html">bionj</a></span>() <span class="kw">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/r/graphics/plot.html">plot</a></span>()</pre></body></html></div>
<p><img src="ref_database_files/figure-html/unnamed-chunk-17-1.png" width="576"></p>
<p>It is important to note that this tree is not an accurate phylogenetic reconstruction but it can help to visualise the hierarchical relationships between the sequences. This is particularly useful with larger datasets to detect polyphyletic clades unresolved by the barcode region.</p>
</div>
<div id="exporting-data" class="section level3">
<h3 class="hasAnchor">
<a href="#exporting-data" class="anchor"></a>Exporting data</h3>
<p>When we are happy with the results, we usually want to export the data. Here, we could export directly the dataframe using <code>write_csv()</code>. However, it might be more convenient to export sequences in FASTA to open them in an other software later. This can be achieved using the function <code><a href="../reference/write_fasta.html">write_fasta()</a></code>.</p>
<div class="sourceCode" id="cb16"><html><body><pre class="r"><span class="no">fra_consensus</span> <span class="kw">%&gt;%</span>
  <span class="fu">select</span>(<span class="no">taxa_group</span>, <span class="no">consensus_barcode</span>) <span class="kw">%&gt;%</span>
  <span class="fu">deframe</span>() <span class="kw">%&gt;%</span>
  <span class="fu"><a href="../reference/write_fasta.html">write_fasta</a></span>(<span class="st">"my_sequences.fasta"</span>)</pre></body></html></div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Francois Keck.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
